libtenzir/src/tql2/parser.cpp

```diff
@@ -488,15 +510,80 @@ class parser {
     };
   }
 
+  auto token_to_str(token_kind kind) -> std::optional<constant> {
+    TENZIR_ASSERT(kind == tk::string || kind == tk::raw_string);
+    const auto token = accept(kind);
+    if (not token) {
+      return std::nullopt;
+    }
+    auto result = std::string{};
+    if (kind == tk::raw_string) {
+      const auto begin = token.text.find('"');
+      const auto end = token.text.rfind('"');
+      TENZIR_ASSERT(begin != std::string_view::npos);
+      TENZIR_ASSERT(end != std::string_view::npos);
+      result = std::string{token.text.substr(begin + 1, end - begin - 1)};
+    } else if (kind == tk::string) {
+      // TODO: Implement this properly.
+      TENZIR_ASSERT(token.text.size() >= 2);
+      TENZIR_ASSERT(token.text.front() == '"');
+      TENZIR_ASSERT(token.text.back() == '"');
+      auto f = token.text.begin() + 1;
+      auto e = token.text.end() - 1;
+      for (auto it = f; it != e; ++it) {
+        auto x = *it;
+        if (x != '\\') {
+          result.push_back(x);
+          continue;
+        }
+        ++it;
+        if (it == e) {
+          // TODO: invalid, but cannot happen
+          TENZIR_UNREACHABLE();
+        }
+        x = *it;
+        if (x == '\\') {
+          result.push_back('\\');
+        } else if (x == '"') {
+          result.push_back('"');
+        } else if (x == 't') {
+          result.push_back('\t');
+        } else if (x == 'n') {
+          result.push_back('\n');
+        } else if (x == '0') {
+          result.push_back('\0');
+        } else {
+          diagnostic::error("found unknown escape sequence `{}`",
+                            token.text.substr(it - f, 2))
+            .primary(token.location.subloc(it - f, 2))
+            .throw_();
+        }
+      }
+    }
+    if (not arrow::util::ValidateUTF8(result)) {
+      // TODO: Would be nice to report the actual error location.
+      diagnostic::error("string contains invalid utf-8")
+        .primary(token)
+        .hint("consider using a blob instead: b{}", token.text)
+        .throw_();
+    }
+    return constant{std::move(result), token.location};
+  }
+
   auto parse_record_item() -> ast::record::item {
     if (accept(tk::dot_dot_dot)) {
       return ast::record::spread{parse_expression()};
     }
     auto name = accept(tk::identifier);
     if (not name) {
       // TODO: Decide how to represent string fields in the AST.
-      name = expect(tk::string);
-      name.text = name.text.substr(1, name.text.size() - 2);
+      auto result = token_to_str(tk::string);
+      if (not result) {
+        result = token_to_str(tk::raw_string);
+      }
+      TENZIR_ASSERT(result);
+      const auto text = std::get<std::string>(result->value);
+      name = accept_result{text, result->get_location()};
```

Comment:
You can't `TENZIR_ASSERT` here that you get a string. The original code did `expect(tk::string)`, which produces a diagnostic in case there is no string.

Also, note the two calls to `token_to_str` and the call to `std::get` here. I'll reference that in my next comment.